DAG: Fix vector bin op scalarize defining a partially undef vector

arsenm opened the pull request:
This avoids some of the pending regressions after AMDGPU implements
isExtractVecEltCheap.

In a case like shl <value, undef>, splat k, because the second operand
was fully defined, we would fall through and use the splat value for the
first operand, losing the undef high bits. This would result in an additional
instruction to handle the high bits. Add some reduced testcases for different
opcodes for one of the regressions.

arsenm committed:
DAG: Fix vector bin op scalarize defining a partially undef vector

This avoids some of the pending regressions after AMDGPU implements
isExtractVecEltCheap.

In a case like shl <value, undef>, splat k, because the second operand
was fully defined, we would fall through and use the splat value for the
first operand, losing the undef high bits. This would result in an additional
instruction to handle the high bits. Add some reduced testcases for different
opcodes for one of the regressions.

arsenm commented:
* **#122460** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/122460?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#122459** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/122459?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/122459?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -27525,8 +27525,12 @@ static SDValue scalarizeBinOpOfSplats(SDNode *N, SelectionDAG &DAG,
+      // TODO: Do we need to check if the opcode/operand propagates undef?
+      // Should we ignore operation identity values?
+      ((count_if(N0->ops(), [](SDValue V) { return !V.isUndef(); }) == 1) ||
+       (count_if(N1->ops(), [](SDValue V) { return !V.isUndef(); }) == 1))) {
```
The opcode specific under handling does worry me- how well does this work if we completely scalarize all elements and leave getNode() to handle the undefs / constant handling etc? Similar to what getKnownUndefForVectorBinop does in TargetLowering

arsenm replied:
I don't actually know why this combine is doing any of this. It works best to skip the whole special handling, and just use INSERT_VECTOR_ELT

arsenm replied:
Actually in the next commit I find a miscompile from this, something is wrong

arsenm replied:
Just doing getNode and checking if the result folded to undef seems to work. I do see some regressions in the test I was looking at, but I think these are actually undef handling bug fixes (e.g. xor undef, undef was previously assumed to return undef but we fold that to 0)

arsenm committed:
Just use insert_vector_elt

arsenm committed:
Check if getNode folded to undef instead

arsenm committed:
Don't bother counting the undefs

It folds to undef or a constant and seems to have the best result
anyway rather than overdefining by splatting the variable value.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -27525,23 +27525,25 @@ static SDValue scalarizeBinOpOfSplats(SDNode *N, SelectionDAG &DAG,
-  SDValue ScalarBO = DAG.getNode(Opcode, DL, EltVT, X, Y, N->getFlags());
-
   // If all lanes but 1 are undefined, no need to splat the scalar result.
   // TODO: Keep track of undefs and use that info in the general case.
```
I think both the comment and TODO need updating

RKSimon replied:
We don't check for all-but-one undef lanes though anymore?

RKSimon reviewed (approved):
LGTM with one minor (update comments)

arsenm committed:
Update comment

arsenm committed:
Fix comment again

arsenm commented:
### Merge activity

* **Jan 16, 8:30 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/122459).
